### ISSUE TITLE
API usage with JCache without caching the proxies on the client side

### ISSUE THREAD

**Issue by jdubois** [vladimir-bukhtoyarov/bucket4j#26]
I'm currently working on https://github.com/jhipster/generator-jhipster/issues/5388

So basically I'm doing a request filter like [here in the documentation](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/basic-usage.md#example-3---limiting-the-rate-of-access-to-rest-api), but using JCache with Hazelcast like [here in the documentation](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/jcache-usage.md#example-of-hazelcast-integration).

Mixing those 2 examples, my current implementation is the following (please note this is a Zuul filter, not a Servlet filter, but this doesn't make any difference):

```Java
String bucketId = getId(RequestContext.getCurrentContext().getRequest());
        
Bucket bucket = Bucket4j.jCacheBuilder(RecoveryStrategy.RECONSTRUCT)
    .addLimit(Bandwidth.simple(jHipsterProperties.getGateway().getRateLimiting().getLimit(),
        Duration.ofSeconds(3_600))
    .build(cache, bucketId);

if (bucket.tryConsumeSingleToken()) {
    // the limit is not exceeded
    log.debug("API rate limit OK for {}", bucketId);
} else {
    // limit is exceeded
    log.info("API rate limit exceeded for {}", bucketId);
    apiLimitExceeded();
}
return null;
```

This code works, and follows the documentation, but it worries me that we "create" a new Bucket for each HTTP request. This looks rather heavy, and I'd better re-use the `Bucket` that was created.

However, we just store a `BucketState` in the cache, and I can't find a way to get a `Bucket` out of the cache.

Am I doing something wrong? I currently think the API should be modified, or maybe better documented, because I can't find a good solution here.

**Comment by vladimir-bukhtoyarov** [comment]
In case of JCache usage scenario bucket divided in two logical parts:
- [GridBucketState](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/src/main/java/com/github/bucket4j/grid/GridBucketState.java) has one-to-one logical relation with bucket and stored in data-grid, most likely you should not worry about this part, JCache will manage (and replicate if configured) for you.
- [GridBucket](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/src/main/java/com/github/bucket4j/grid/GridBucket.java) acts like proxy and just issues commands for bucket mutation, then commands serialized and executed on the JCache node which holds the GridBucketState instance. In opposite to GridBucketState, GridBucket has many-to-one relation with bucket, because bucket can be accessible from multiple client node, and even multiple times on same(as in your example) if client has no rational strategy to cache the instance of GridBucket.

GridBucket can be created multiple time and this will not lead to logical error, because the main part of bucket GridBucketState is strongly protected from duplication, but this will lead to significant performance degradation, because GridBucket [issues network request](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/src/main/java/com/github/bucket4j/grid/GridBucket.java#L34) to cache at the moment of bucket initialization, so you will pay at least twice for non-optimized code. So, **GridBucket must not be treated as lightweit entity**, it is better to cache its instance and reuse between invocations.

Unfortunately caching the GridBucket instances for future reusing is your responsibility, I argue thet it is wrong to introduce this kind of functionality at the library level. In the JEE environment you can just use servlet session for beginning, [see this](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/basic-usage.md#example-3---limiting-the-rate-of-access-to-rest-api) as example. If do not trust to servlet container(or there is no container) then you can just manage bucket proxies in any cache like Guava or Caffeine, and moreover you can limit the lifetime of instances(because rarely losing of proxie will not lead to error) by size, last access time and any other meaningful limits specific for your application profile. So as you can see, there are many ways to caching strong reference to proxy, and that it why I do not want to involve any of them at library level.

According to documentation, I agree that several points from this discussion should be added to wiki and javadocs. I plan to enhance documentation in scope of release 2.0

**Comment by vladimir-bukhtoyarov** [comment]
Of coarse it is not a rocket science to make JCache implementation of token-bucket without requirements to caching the proxies, but API will be ugly(@yegor256 bit me), because in case of instance-less API, all consumption method must be static and looks like this:
```java
static boolean tryConsume(long tokens, Cache cache, String bucketKey, Supplier<Configuration> configSupplier) {
    // configSupplier invoked if and only if bucket does not present in cache
    ...
}
```
And instance-less JCache API will be incompatible with API for local bucket, hello to double work related to implementation, benchmarking, documentation and testing. So I will never provide instance-less API version, because caching of proxies does not look for me as very complex task.

**Comment by jdubois** [comment]
Thanks for the quick response @vladimir-bukhtoyarov - and this indeed confirms what I thought. However I'm in a distributed environnement, so I can't store the proxies in a local cache - I need to think about this, I'll post an update if I have something good.

**Comment by vladimir-bukhtoyarov** [comment]
> However I'm in a distributed environnement, so I can't store the proxies in a local cache

Does not matter how many proxies to same bucket exists in same time, existence several proxies to same bucket is not a problem. Proxy is very stuppid, do not care about theme to much. The core of bucket is GridBucketState, bucket4j transparently provides guarantee that only one state will present for same create. So you are free to create as many proxies as you need, just avoid to create it per single invocation basis.

**Comment by vladimir-bukhtoyarov** [comment]
> please note this is a Zuul filter, not a Servlet filter

I missed that yesterday. I done a quick investigation of your [RateLimitingFilter](https://github.com/jhipster/generator-jhipster/blob/da5b9c83b3fcc1a9587938a544a250524740e740/generators/server/templates/src/main/java/package/gateway/ratelimiting/_RateLimitingFilter.java#L2) and I confirm that better solution will be as described above:
> just manage bucket proxies in any cache like Guava or Caffeine.

I [found the pretty way](https://github.com/vladimir-bukhtoyarov/bucket4j/issues/27) to keep existing (Bucket)[https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/src/main/java/com/github/bucket4j/Bucket.java] interface for JCache and provide ability to acts without caching of proxy instances on user side, but this task proposed to be done in Jule 2017. So if you want to replace Cassandra asap then local cache is only valid solution. And of coarse, auto detection of split-brain and other datagrid related issues will be impossible in new style API, because without of caching proxy it is impossible to determine that persisted part of bucket was lost, so old style API for JCache will be supported as well, at least for users which highly interested for consistency and problem detection

**Comment by jdubois** [comment]
Thanks @vladimir-bukhtoyarov - indeed the "instance less API" wouldn't have the same consistency: missing one or two HTTP requests isn't very important for rate limiting (at least for normal use cases), so this solution would be much better for me.

I'll do some more research, but if I don't find a better solution, I'll wait for July 2017!

**Comment by vladimir-bukhtoyarov** [comment]
Ok, @jdubois let me know if found something interest. 

Also I want to notice that you are configured only one limit, and in same time, the period of limitation is very long(1 hour if I understand correctly), it is very dangerous, I'll try to explain why single limitation with long period is bad:
- Given the limitation 10000 tokens/ per 1 hour. As a hacker I am able to send 9999 request in very short period, for example at first 10 seconds, and potentially 1000RPS can damage or slowdown any part of your service.
- Then as good hacker I stop to sending, and bucket is unable to catch me, because limit was not reached.
- Then I will be reproduce this attack each hour.

To protect from this kind attacks, Bucket4j provides two different options:
- You can specify multiple limits [as describe there](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/advanced-usage.md#example-of-multiple-bandwidth).
```java
Bucket bucket = Bucket4j.jCacheBuilder(RecoveryStrategy.RECONSTRUCT)
    .addLimit(Bandwidth.simple(10000, Duration.ofSeconds(3_600))
    .addLimit(Bandwidth.simple(20, Duration.ofSeconds(1)) // hacker is unable to achieve 1000RPS and crash service in short time
    .build(cache, bucketId);
```
From performance perspective it does not matter how many limits specified per single bucket.
- There is another way to protect form storming - just forget about "period" of storming in think in term of "overdraft" [as described there](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/basic-usage.md#example-3---limiting-the-rate-of-access-to-rest-api). The "overdraft" feature was originally asked by @efenderbosch in #13, and Bucket4j supports this feature since release 1.3.

The first option is very easy to understand, thinking in term of overdraft is more complex than using multiple bandwidths, but second way is [originally invented by authors](https://en.wikipedia.org/wiki/Token_bucket#Algorithm) of token-bucket algorithm, so many old-school guys can prefer the second way. And of course you can combine first and second options as you wish to achieve non-compromise protection.

So, during searching for alternatives keep in mind that currently your throttling solution based on the top of Cassandra is not protected from short time storming.

**Comment by jdubois** [comment]
Thanks @vladimir-bukhtoyarov I already know those issues: I'm just doing a very simple and naive implementation, then it's up to people using JHipster to create better limits, according to their specific business needs.
I might do something a bit more complex with Bucket4J (which I couldn't do with my current Cassandra-based implementation), but at the moment my main issue is making this work when I scale my gateways up.

**Comment by vladimir-bukhtoyarov** [comment]
it was easy than planned and took around 2 hours.

@jdubois could you provide feedback according to [this commit](https://github.com/vladimir-bukhtoyarov/bucket4j/commit/0bd21548d08191ab48feb2527902c8bd942a66f9)?

The key points to review:
- [createLazyBucket](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/2.0/bucket4j-core/src/main/java/io/github/bucket4j/grid/GridBucket.java#L34) creates very cheap proxy to bucket, without initialization at construction time.
- [JCacheBucketRegistry](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/2.0/bucket4j-core/src/main/java/io/github/bucket4j/grid/jcache/JCacheBucketRegistry.java) is abstraction for collection of lazy buckets.
- [HazelcastTest#testJCacheBucketRegistryWithKeyIndependentConfiguration](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/2.0/bucket4j-core/src/test/java/io/github/bucket4j/grid/jcache/hazelcast/HazelcastTest.java#L78) example which shows how to use JCacheBucketRegistry if all buckets have same configuration.
- [HazelcastTest#testJCacheBucketRegistryWithKeyDependentConfiguration](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/2.0/bucket4j-core/src/test/java/io/github/bucket4j/grid/jcache/hazelcast/HazelcastTest.java#L94) example which shows how to use JCacheBucketRegistry if all buckets have same configuration.

Most likely I will release version "2.0" at next Sunday(it is need to complete javadocs and wiki pages), currently you are able to build library from source.

**Comment by jdubois** [comment]
Thank you so much! However I won't be able to work on this until next week, sorry.

**Comment by vladimir-bukhtoyarov** [comment]
Done, will be released in scope of "2.0".

**Comment by stevehouel** [comment]
Thx @vladimir-bukhtoyarov  for those changes in v2. 

@jdubois need some help ?
